exporter/signalfxexporter/translation/converter.go

```diff
@@ -94,287 +83,273 @@ func NewMetricsConverter(logger *zap.Logger, t *MetricTranslator) *MetricsConver
+			continue
+		}
+		for k := 0; k < ilm.Metrics().Len(); k++ {
+			m := ilm.Metrics().At(k)
+			if m.IsNil() {
+				continue
+			}
 
-	extraDimensions = appendResourceAttributesToDimensions(extraDimensions, resourceAttribs)
-	extraDimensions = appendAttributesToDimensions(extraDimensions, nodeAttribs)
+			dps, droppedDPCount := c.metricToSfxDataPoints(m, res)
+			numDroppedTimeSeries += droppedDPCount
 
-	for _, metric := range md.Metrics {
-		if metric == nil || metric.MetricDescriptor == nil {
-			c.logger.Warn("Received nil metrics data or nil descriptor for metrics")
-			numDroppedTimeSeries += len(metric.GetTimeseries())
-			continue
+			sfxDatapoints = append(sfxDatapoints, dps...)
 		}
+	}
+	sanitizeDataPointDimensions(sfxDatapoints)
+	return sfxDatapoints, numDroppedTimeSeries
+}
 
-		metricDataPoints := make([]*sfxpb.DataPoint, 0, len(metric.Timeseries))
+func (c *MetricsConverter) metricToSfxDataPoints(metric pdata.Metric, res pdata.Resource) ([]*sfxpb.DataPoint, int) {
+	if metric.IsNil() {
+		return nil, 0
+	}
 
-		// Build the fixed parts for this metrics from the descriptor.
-		descriptor := metric.MetricDescriptor
-		metricName := descriptor.Name
+	var extraDimensions []*sfxpb.Dimension
+	if !res.IsNil() {
+		resourceAttribs := res.Attributes()
+		extraDimensions = make([]*sfxpb.Dimension, 0, resourceAttribs.Len())
+		extraDimensions = appendResourceAttributesToDimensions(extraDimensions, resourceAttribs)
+	}
 
-		metricType := fromOCMetricDescriptorToMetricType(descriptor.Type)
-		numLabels := len(descriptor.LabelKeys)
+	var dps []*sfxpb.DataPoint
+
+	basePoint := makeBaseDataPoint(metric)
+
+	switch metric.DataType() {
+	case pdata.MetricDataTypeNone:
+		return nil, 0
+	case pdata.MetricDataTypeIntGauge:
+		dps = convertIntDatapoints(metric.IntGauge().DataPoints(), basePoint, extraDimensions)
+	case pdata.MetricDataTypeIntSum:
+		dps = convertIntDatapoints(metric.IntSum().DataPoints(), basePoint, extraDimensions)
+	case pdata.MetricDataTypeDoubleGauge:
+		dps = convertDoubleDatapoints(metric.DoubleGauge().DataPoints(), basePoint, extraDimensions)
+	case pdata.MetricDataTypeDoubleSum:
+		dps = convertDoubleDatapoints(metric.DoubleSum().DataPoints(), basePoint, extraDimensions)
+	case pdata.MetricDataTypeIntHistogram:
+		dps = convertIntHistogram(metric.IntHistogram().DataPoints(), basePoint, extraDimensions)
+	case pdata.MetricDataTypeDoubleHistogram:
+		dps = convertDoubleHistogram(metric.DoubleHistogram().DataPoints(), basePoint, extraDimensions)
+	}
 
-		for _, series := range metric.Timeseries {
-			dimensions := make([]*sfxpb.Dimension, numLabels+len(extraDimensions))
-			copy(dimensions, extraDimensions)
-			for i := 0; i < numLabels; i++ {
-				dimension := &sfxpb.Dimension{
-					Key:   descriptor.LabelKeys[i].Key,
-					Value: series.LabelValues[i].Value,
-				}
-				dimensions[len(extraDimensions)+i] = dimension
-			}
+	if c.metricTranslator != nil {
+		dps = c.metricTranslator.TranslateDataPoints(c.logger, dps)
+	}
```

Comment:
Does the translator change dimensions in-place (modifies a dimension that we pass as pointer directly)? I am asking because if that is the case we need to do a deeper copy of the extra dimensions?

Reply:
See https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/1375#discussion_r513438802